File: viewer/src/main/scala/farjs/viewer/ViewerPlugin.scala
package farjs.viewer

import farjs.file.FileEvent.onFileView
import farjs.file.FileViewHistory
import farjs.filelist._
import farjs.filelist.api.FileListItem
import farjs.filelist.stack.PanelStacks
import farjs.viewer.ViewerEvent._
import scommons.nodejs._
import scommons.react._

import scala.scalajs.js

object ViewerPlugin extends FileListPlugin(js.Array(
  "f3", onViewerOpenLeft, onViewerOpenRight, onFileView
)) {

  private[viewer] final var fs: FS = scommons.nodejs.fs

  override def onKeyTrigger(key: String,
                            stacks: PanelStacks,
                            data: js.UndefOr[js.Dynamic] = js.undefined
                           ): js.Promise[js.UndefOr[ReactClass]] = new js.Promise[js.UndefOr[ReactClass]]((resolve, _) => {
    
    val res =
      if (key == onFileView) {
        data.toOption.map { d =>
          val history = d.asInstanceOf[FileViewHistory]
          val size = fs.lstatSync(history.path).size
          new ViewerPluginUi(history.path, size).apply()
        }
      }
      else {
        val stack = key match {
          case `onViewerOpenLeft` => stacks.left.stack
          case `onViewerOpenRight` => stacks.right.stack
          case _ => PanelStacks.active(stacks).stack
        }
        val stackItem = stack.peek[FileListState]
        stackItem.getData().toOption.flatMap { data =>
          val FileListData(_, actions, state) = data
          FileListState.currentItem(state).filter(_ != FileListItem.up).toOption match {
            case Some(item) if actions.api.isLocal && !item.isDir =>
              val filePath = path.join(state.currDir.path, item.name)
              val size = fs.lstatSync(filePath).size
              val ui = new ViewerPluginUi(filePath, size)
              Some(ui.apply())
            case Some(item) if state.selectedNames.nonEmpty || item.isDir =>
              Some(new ViewItemsPopup(data).apply())
            case _ => None
          }
        }
      }

    resolve(res match {
      case Some(r) => r
      case None => js.undefined
    })
  })
}

File: viewer/ViewerPlugin.mjs
/**
 * @typedef {import("@farjs/filelist/stack/WithStacksProps.mjs").WithStacksProps} WithStacksProps
 * @import { ReactComponent } from "@farjs/filelist/FileListData.mjs"
 * @import { FileViewHistory } from "../file/FileViewHistory.mjs"
 */
import fs from "fs";
import path from "path";
import FileListItem from "@farjs/filelist/api/FileListItem.mjs";
import WithStacksProps from "@farjs/filelist/stack/WithStacksProps.mjs";
import PanelStackItem from "@farjs/filelist/stack/PanelStackItem.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";
import FileListPlugin from "@farjs/filelist/FileListPlugin.mjs";
import FileEvent from "../file/FileEvent.mjs";
import ViewerEvent from "./ViewerEvent.mjs";
import ViewItemsPopup from "./ViewItemsPopup.mjs";
import ViewerPluginUi from "./ViewerPluginUi.mjs";

class ViewerPluginImpl extends FileListPlugin {
  constructor() {
    super([
      "f3",
      ViewerEvent.onViewerOpenLeft,
      ViewerEvent.onViewerOpenRight,
      FileEvent.onFileView,
    ]);
  }

  /**
   * @param {string} key
   * @param {WithStacksProps} stacks
   * @param {any} [data]
   * @returns {Promise<ReactComponent | undefined>}
   */
  async onKeyTrigger(key, stacks, data) {
    if (key === FileEvent.onFileView) {
      if (data) {
        /** @type {FileViewHistory} */
        const history = data;
        const size = fs.lstatSync(history.path).size;
        return ViewerPluginUi(history.path, size);
      }

      return undefined;
    }

    const stack = (() => {
      switch (key) {
        case ViewerEvent.onViewerOpenLeft:
          return stacks.left.stack;
        case ViewerEvent.onViewerOpenRight:
          return stacks.right.stack;
        default:
          return WithStacksProps.active(stacks).stack;
      }
    })();

    /** @type {PanelStackItem<FileListState>} */
    const stackItem = stack.peek();
    const fileListData = stackItem.getData();
    if (fileListData) {
      const { actions, state } = fileListData;
      const item = FileListState.currentItem(
        state,
        (_) => _ !== FileListItem.up
      );

      if (item && actions.api.isLocal && !item.isDir) {
        const filePath = path.join(state.currDir.path, item.name);
        const size = fs.lstatSync(filePath).size;
        return ViewerPluginUi(filePath, size);
      }

      if (state.selectedNames.size > 0 || (item && item.isDir)) {
        return ViewItemsPopup(fileListData);
      }
    }

    return undefined;
  }
}

const ViewerPlugin = new ViewerPluginImpl();

export default ViewerPlugin;
